### PR TITLE
Adding CA cert functionality to elastic

### PR DESCRIPTION
**Pull request by bmacnee**
https://github.com/elastic/apm-agent-go/issues/752

Attempting to add functionality similar to other elastic-apm agents where a ca cert can be provided instead of the literal server cert.

I am a bit confused by the comment `TestHTTPTransportServerCert` :
```	
// Reconfigure the transport so that it knows about the
// CA certificate. We avoid using server.Client here, as
// it is not available in older versions of Go.
```
	
This refers to a CA certificate, but my understanding is this functionality would only work for the servers cert.

### CODE REVIEW

**Comment by bmacnee** [comment]
signed the Agreement after the fact, dunno if there's a way to kick off the githooks again...

File: transport/http.go [hunk 1]
```diff
@@ -144,6 +145,22 @@ func NewHTTPTransport() (*HTTPTransport, error) {
 		}
 	}
 
+	caCertPath := os.Getenv(envCACert)
+	if caCertPath != "" {
+		rootCAs, err := x509.SystemCertPool()
+		if err != nil || rootCAs == nil {
+			rootCAs = x509.NewCertPool()
+		}
```
Comment (axw) [inline transport/http.go:153]: I think if you configure a custom CA cert, we should not be using the system cert pool at all.

```suggestion
		rootCAs := x509.NewCertPool()
```

File: transport/http.go [hunk 2]
```diff
@@ -144,6 +145,22 @@ func NewHTTPTransport() (*HTTPTransport, error) {
...
+		additionalCerts, err := ioutil.ReadFile(caCertPath)
+		if err != nil {
+			return nil, errors.Wrapf(err, "failed to load root CA file from %s", caCertPath)
+		}
+		// TODO: could return an error or a log if no additional certs are loaded
+		rootCAs.AppendCertsFromPEM(additionalCerts)
```
Comment (axw) [inline transport/http.go:159]: yes please :)

```suggestion
		if !rootCAs.AppendCertsFromPEM(additionalCerts) {
			return nil, fmt.Errorf("failed to load CA certs from %s", caCertPath)
		}
```

should be good enough I think?

File: transport/http_test.go [hunk 3]
```diff
@@ -314,6 +314,51 @@ func TestHTTPTransportV2NotFound(t *testing.T) {
 	assert.EqualError(t, err, fmt.Sprintf("request failed with 404 Not Found: %s/intake/v2/events not found (requires APM Server 6.5.0 or newer)", server.URL))
 }
 
+func TestHTTPCACert(t *testing.T) {
```
Comment (axw) [inline transport/http_test.go:317]:
```suggestion
func TestHTTPTransportCACert(t *testing.T) {
```

for consistency

**Review by axw — CHANGES_REQUESTED** [review]
This is great, thanks for your contribution! Just a few small comments, otherwise LGTM.

> This refers to a CA certificate, but my understanding is this functionality would only work for the servers cert.

Yes indeed -- sorry, the comment is wrong :sweat_smile:
While you're in the vicinity, would you mind replacing "CA" with "server" in that comment?

File: transport/http.go (hunk 1 above)
Comment (bmacnee) [reply transport/http.go:153]: cool by me! will change.

**Comment by bmacnee** [comment]
@axw curious on a ballpark estimate for how long it will take you folks to release the agent with these changes, as the work I'm doing is dependent on them. Thank you!! Worst case I can use my branch if it will take some time.

**Review by axw — APPROVED** [review]
> @axw curious on a ballpark estimate for how long it will take you folks to release the agent with these changes, as the work I'm doing is dependent on them. Thank you!! Worst case I can use my branch if it will take some time.

@bmacnee there's nothing specifically scheduled, but I anticipate we'll cut a new release in 3-4 weeks.

**Comment by axw** [comment]
jenkins run the tests please

**Comment by axw** [comment]
Go 1.8/1.9 failures are unrelated, will be fixed by https://github.com/elastic/apm-agent-go/pull/984 -- merging.

Thanks again!

**Pull request merged**